storage/gcp/gcp.go

```diff
@@ -221,11 +225,92 @@ type Appender struct {
 	queue *storage.Queue
 
 	cpUpdated chan struct{}
+
+	// This mutex guards the stopped state. We use this instead of an atomic.Boolean
+	// to get the property that no readers of this state can have the lock when the
+	// write gets it. This means that no in-flight Add operations will be occurring on
+	// Shutdown.
+	mu      sync.RWMutex
+	stopped bool
+
+	// This channel will behave as per Context.Done() to signify that this appender is finished.
+	done <-chan struct{}
+
+	// largestIssued tracks the largest index allocated by this appender.
+	largestIssued atomic.Uint64
 }
 
 // Add is the entrypoint for adding entries to a sequencing log.
 func (a *Appender) Add(ctx context.Context, e *tessera.Entry) tessera.IndexFuture {
-	return a.queue.Add(ctx, e)
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	if a.stopped {
+		return func() (uint64, error) {
+			return 0, errors.New("appender has been shut down")
+		}
+	}
+	res := a.queue.Add(ctx, e)
+	return func() (uint64, error) {
+		i, err := res()
+		if err != nil {
+			return i, err
+		}
+
+		// https://github.com/golang/go/issues/63999 - atomically set largest issued index
+		old := a.largestIssued.Load()
+		for old < i && !a.largestIssued.CompareAndSwap(old, i) {
+			old = a.largestIssued.Load()
+		}
+
+		return i, err
+	}
+}
+
+// Shutdown ensures that all calls to Add that have returned a value will be resolved. Any
+// futures returned by _this appender_ which resolve to an index will be integrated and have
+// a checkpoint that commits to them published if this returns successfully.
+//
+// After this returns, any calls to Add will fail.
+func (a *Appender) Shutdown(ctx context.Context) error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.stopped = true
+	if err := a.queue.Close(ctx); err != nil {
+		return err
+	}
+	// At this point sequencing and integration is done, now to make sure a large
+	// enough checkpoint has been issued.
+	maxIndex := a.largestIssued.Load()
+	if maxIndex == 0 {
+		// special case no work done
+		close(a.cpUpdated)
+		<-a.done
```

Comment:
You're racing the `publisherTask` goroutine in taking from the `cpUpdated` channel here, so the special case could occur over there (I believe the Go runtime can do context switching to another goroutine on function calls).

I think the worst that could happen if you just let it handle it there is that it'll publish one more CP and then exit due to the closed `cpUpdated` channel?

I imagine the same in the other storage impls.

Reply:
There's a comment on the channel field that we pull from. Discussed 1:1 and you are happy that's reasonable.